einops/einops.py

```diff
@@ -323,6 +323,47 @@ def _prepare_transformation_recipe(pattern: str,
         ellipsis_position_in_lhs=ellipsis_left,
     )
 
+@functools.lru_cache(256)
+def _match_einop(pattern: str, reduction=None, **axes_lengths: int):
+    """Find the corresponding operation matching the pattern"""
+    left, rght = pattern.split('->')
+    left = ParsedExpression(left)
+    rght = ParsedExpression(rght)
+
+    default_op = 'rearrange'
+    op = default_op
+
+    for index in left.identifiers:
+        if index not in rght.identifiers:
+            op = 'reduce'
+            break
+
+    for index in rght.identifiers:
+        if index not in left.identifiers:
+            if op != default_op:
+                raise EinopsError('You must perform a reduce and repeat separately: {}'.format(pattern))
+            op = 'repeat'
+            break
+
+    return op
+
+def einop(tensor, pattern: str, reduction=None, **axes_lengths: int):
+    """Perform either reduce, rearrange, or repeat depending on pattern"""
+    op = _match_einop(pattern, reduction, **axes_lengths)
+
+    if op == 'rearrange':
+        if reduction is not None:
+            raise EinopsError('Do not pass reduction for rearrange pattern: {}'.format(pattern))
```

Comment:
I think this error message is "leaking" implementation details (the use of rearrange). Consider something like:

> Got reduction operation but there is no dimension to reduce in pattern: '{pattern}'

Reply:
Thanks, fixed.